tools/sotn-assets/graphics.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"encoding/binary"
+	"os"
+)
+
+type gfxKind uint16
+
+const (
+	gfxBankNone = gfxKind(iota)
+	gfxBank4bpp
+	gfxBank8bpp
+	gfxBank16bpp
+	gfxBankCompressed
+)
+
+type gfxEntry struct {
+	X      uint16
+	Y      uint16
+	Width  uint16
+	Height uint16
+	Data   PsxOffset
+}
+
+type gfxBlock struct {
+	kind    gfxKind
+	flags   uint16
+	entries []gfxEntry
+}
+
+type gfx struct {
+	blocks  []gfxBlock
+	indices []int
+}
+
+func readGraphics(file *os.File, off PsxOffset) (gfx, dataRange, error) {
+	if err := off.moveFile(file); err != nil {
+		return gfx{}, dataRange{}, err
+	}
+
+	// all the offsets are before the array, so it is easy to find where the offsets array ends
+	blockOffsets := []PsxOffset{}
+	for {
+		var offBank PsxOffset
+		if err := binary.Read(file, binary.LittleEndian, &offBank); err != nil {
+			return gfx{}, dataRange{}, err
+		}
+		if offBank >= off {
+			break
+		}
+		blockOffsets = append(blockOffsets, offBank)
+	}
+
+	// the order of each gfxBlock must be preserved
+	pool := map[PsxOffset]int{}
+	pool[RamNull] = -1
+	blocks := []gfxBlock{}
+	ranges := []dataRange{}
+	for _, blockOffset := range sortUniqueOffsets(blockOffsets) {
+		if blockOffset == RamNull { // exception for ST0
+			continue
+		}
+		if err := blockOffset.moveFile(file); err != nil {
+			return gfx{}, dataRange{}, err
+		}
+		var block gfxBlock
+		if err := binary.Read(file, binary.LittleEndian, &block.kind); err != nil {
+			return gfx{}, dataRange{}, err
+		}
+		if err := binary.Read(file, binary.LittleEndian, &block.flags); err != nil {
+			return gfx{}, dataRange{}, err
+		}
+
+		if block.kind == gfxKind(0xFFFF) && block.flags == 0xFFFF { // exception for ST0
```

Comment:
maybe a more detailed comment explaining what the exception is and why it's required?

Reply:
The graphics stuff is still experimental and I am not yet extracting it into files as part of this PR because some assumptions are wrong. I probably need to rewrite that part anyway.